hax/hax/handler.py

```diff
@@ -107,45 +106,39 @@ def update_process_failure(self, q: Queue,
                     # the real time status of the process before
                     # broadcasting failure.
                     current_status = ServiceHealth.UNKNOWN
-                    q.put(BroadcastHAStates(
-                          states=[HAState(fid=state.fid,
-                                          status=ServiceHealth.FAILED)],
-                          reply_to=None))
-                new_ha_states.append(HAState(fid=state.fid,
-                                             status=current_status))
+                    planner.add_command(
+                        BroadcastHAStates(states=[
+                            HAState(fid=state.fid, status=ServiceHealth.FAILED)
+                        ],
+                                          reply_to=None))
+                new_ha_states.append(
+                    HAState(fid=state.fid, status=current_status))
             else:
                 new_ha_states.append(state)
         return new_ha_states
 
-    def _do_work(self, q: Queue, motr: Motr):
+    def _do_work(self, planner: WorkPlanner, motr: Motr):
         ffi = motr._ffi
         LOG.info('Handler thread has started')
         ffi.adopt_motr_thread()
 
-        def pull_msg():
-            try:
-                return q.get(block=False)
-            except Empty:
-                return None
-
         try:
             while True:
                 try:
                     LOG.debug('Waiting for the next message')
 
-                    item = pull_msg()
-                    while item is None:
-                        time.sleep(0.2)
-                        if self.is_stopped:
-                            raise StopIteration()
-                        item = pull_msg()
+                    item = planner.get_next_command()
+                    # TODO maybe it makes sense to move ensure_allowed() into
+                    # get_next_command()?
+                    item = planner.ensure_allowed(item)
```

Comment:
Yes, I think it would be good to merge get_next_command() and ensure_allowed() so the the user of get_next_command() need not have to have knowledge about one more step. The command returned by get_next_command() can be executed directly.

Reply:
@mssawant I've addressed the latest comments and rebased the PR against the latest main branch (there were conflicts, hopefully I resolved them properly).

Please have a look.